vowpalwabbit/cb_explore_adf_bag.cc

```diff
@@ -81,10 +81,10 @@ void cb_explore_adf_bag::predict_or_learn(
     cb_explore_adf_bag& data, LEARNER::multi_learner& base, multi_ex& examples)
 {
   if (is_learn)
-    cb_explore_adf_base::learn(data, &cb_explore_adf_bag::predict_or_learn_impl<true>,
+    data.learn(data, &cb_explore_adf_bag::predict_or_learn_impl<true>,
```

Comment:
Is it possible to hide the fact we need to pass the function pointers into the class implementation of learn? So then we can call the function like:
```C++
data.learn(base, examples);
```

Reply:
No, you can't even pass in the non-qualified function name like
```
data.learn(data, &predict_or_learn_impl<true>, ...)
```
The template requires a fully qualified function name, otherwise it gets confused. Its one of the joys of working with member function pointers.

The only way to get away with calling data.learn(...) would be to have the learn_or_predict_impl function be a virtual function

Reply:
What if the constructor of the child passed the function pointers to the base class?

Reply:
I feel like there would be a chicken/egg problem there. The pointer type has the object type encoded within it, so the base class itself would need to be templatized with the derived object's type

Reply:
That's CRTP right?

Reply:
Sure, with CRTP I think I can actually get rid of the entire function pointer stuff. One downside is that the entire class would have to be defined in the header.

It opens up a couple of different ways to implement this behavior, but there are pros and cons to each way. Is simplifying the calls a bit worth a hard-to-read base header?

Reply:
Yeah perhaps it isn't. I'm mainly thinking from the perspective of, if we were to instantiate and use this class on its own, how would it feel?

Reply:
The only publicly exposed interface into the class is the static predict_or_learn, which will do the setup for you. The only burden is on somebody writing the reduction. Again, that can be offloaded to the base class instead (which might be better, i'm open to opinions), but won't leak to external callers

Reply:
Ah okay, I see. So that static function hides the work required to feed the predict function down. Sounds good to me

Reply:
Refactored for a cleaner interface. The base type now how the static predict/learn/finish_multiline_example functions and no function pointers are passed around

The explore type object just needs to implement a public predict/learn function of the appropriate signature